### ISSUE TITLE
[v1.2.1] logger.Post blocks forever when the fluentd server has a certain trouble

### ISSUE THREAD

**Issue by yoheimuta** [fluent/fluent-logger-golang#44]
I encountered a problem that apps using fluent-logger-golang failed to process requests due to timeout while the fluentd didn't output any error log.  See below in detail.

It is certain that server errors are the root cause of this problem. However, I would like to take some measures on the client side for the future.

How about implementing to be enable to set a timeout to logger.Post ? If it's reasonable, I can create a PR later.

# REPRODUCED MANUAL:

Installed version

- fluentd
  - 0.12.31
- fluent-plugin-forest
  - 0.3.3
  - When using a previous version before merging https://github.com/tagomoris/fluent-plugin-forest/pull/28, fluentd doesn't output any error log.
- fluent-plugin-bigquery
  - 0.3.4
- google-api-client
  - 0.9.23
  - ref. https://github.com/google/google-api-ruby-client/issues/521#issuecomment-274379145

<details>
<summary>td-agent.rb</summary>

```rb
source {
  id 'in-forward'
  type :forward
  port 24224
}

##
# in app
[
  "service",
].each do |app|
  match("hogehoge.#{app}.*") {
    id "forest-hogehoge#{app}"
    type :forest
    subtype :copy
    template {
      id "copy-hogehoge#{app}"
      ##
      # BigQuery
      store {
        id "out-bigquery-hogehoge#{app}"
        log_level 'warn'
        type :bigquery
        method :insert
        auth_method :private_key
        email 'hogehoge'
        private_key_path '/etc/td-agent/hogehoge-547f3ca2f35c.p12'
        project 'turnkey-conduit-708'
        dataset :hogehoge
        table "#{app}"
        fetch_schema true
        flush_interval '3s'
        try_flush_interval 0.05
        queued_chunk_flush_interval 0.01
        buffer_type :file
        buffer_chunk_limit '512k'
        buffer_queue_limit 1000
        buffer_chunk_records_limit 300
        retry_limit 10
        max_retry_wait '1s'
        num_threads 10
        buffer_path '/hogehoge/tmp/td-agent/bq-buffer/${tag}'
      }
    }
  }
end
```
</details>

<details>
<summary>main.go</summary>

```go
package main

import (
        "fmt"
        "log"

        "github.com/fluent/fluent-logger-golang/fluent"
)

func main() {
        logger, err := fluent.New(fluent.Config{FluentPort: 24224, FluentHost: "127.0.0.1"})
        if err != nil {
                fmt.Println(err)
        }
        defer logger.Close()
        tag := "hogehoge.service.test"
        var data = map[string]string{
                "foo":  "bar",
                "hoge": "hoge"}
        for i := 0; i < 1000000; i++ {
                e := logger.Post(tag, data)
                if e != nil {
                        log.Println("Error while posting log: ", e)
                } else if (i % 10000) == 0{
                        log.Println("Success to post log: ", i)
                }
        }
}
```
</details>

<details>
<summary>go run main.go</summary>

```
$ go run main.go
2017/01/26 18:14:40 Success to post log:  0
2017/01/26 18:14:40 Success to post log:  10000
2017/01/26 18:14:40 Success to post log:  20000
2017/01/26 18:14:40 Success to post log:  30000
2017/01/26 18:14:40 Success to post log:  40000
2017/01/26 18:14:40 Success to post log:  50000
2017/01/26 18:14:40 Success to post log:  60000
2017/01/26 18:14:40 Success to post log:  70000
2017/01/26 18:14:40 Success to post log:  80000
2017/01/26 18:14:40 Success to post log:  90000
2017/01/26 18:14:40 Success to post log:  100000
2017/01/26 18:14:40 Success to post log:  110000
# It kept blocking even after 10 minutes or more.
```
</details>

# ONE SOLUTION:

When I fixed it to set a timeout of 10 seconds, the logger started to reconnect regularly. As a result, the behavior of blocking for a long time does not occur and the logger reports a certain problem.

- ref. https://golang.org/pkg/net/#TCPConn.SetWriteDeadline

<details>
<summary>fluent/fluent.go</summary>

```diff
diff --git a/fluent/fluent.go b/fluent/fluent.go
index 655f623..30db90b 100644
--- a/fluent/fluent.go
+++ b/fluent/fluent.go
@@ -4,7 +4,6 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
-	"io"
 	"math"
 	"net"
 	"reflect"
@@ -46,7 +45,7 @@ type Fluent struct {
 	pending []byte
 
 	muconn       sync.Mutex
-	conn         io.WriteCloser
+	conn         net.Conn
 	reconnecting bool
 }
 
@@ -297,6 +296,7 @@ func (f *Fluent) send() error {
 
 	var err error
 	if len(f.pending) > 0 {
+		f.conn.SetWriteDeadline(time.Now().Add(10 * time.Second))
 		_, err = f.conn.Write(f.pending)
 		if err != nil {
 			f.conn.Close()
```
</details>

<details>
<summary>go run main.go</summary>

```
[18:16:09 mtburn@ip-10-0-7-194 dev /(T_T)\]$ go run main.go
2017/01/26 18:16:20 Success to post log:  0
2017/01/26 18:16:20 Success to post log:  10000
2017/01/26 18:16:20 Success to post log:  20000
2017/01/26 18:16:20 Success to post log:  30000
2017/01/26 18:16:20 Success to post log:  40000
2017/01/26 18:16:20 Success to post log:  50000
2017/01/26 18:16:20 Success to post log:  60000
2017/01/26 18:16:20 Success to post log:  70000
2017/01/26 18:16:20 Success to post log:  80000
2017/01/26 18:16:20 Success to post log:  90000
2017/01/26 18:16:20 Success to post log:  100000
2017/01/26 18:16:20 Success to post log:  110000
2017/01/26 18:16:30 Error while posting log:  write tcp 127.0.0.1:51946->127.0.0.1:24224: i/o timeout
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:30 Success to post log:  120000
2017/01/26 18:16:30 Success to post log:  130000
2017/01/26 18:16:30 Success to post log:  140000
2017/01/26 18:16:30 Success to post log:  150000
2017/01/26 18:16:30 Success to post log:  160000
2017/01/26 18:16:30 Success to post log:  170000
2017/01/26 18:16:30 Success to post log:  180000
2017/01/26 18:16:30 Success to post log:  190000
2017/01/26 18:16:31 Success to post log:  200000
2017/01/26 18:16:31 Success to post log:  210000
2017/01/26 18:16:31 Success to post log:  220000
2017/01/26 18:16:31 Success to post log:  230000
2017/01/26 18:16:41 Error while posting log:  write tcp 127.0.0.1:51950->127.0.0.1:24224: i/o timeout
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:41 Success to post log:  240000
2017/01/26 18:16:41 Success to post log:  250000
2017/01/26 18:16:41 Success to post log:  260000
2017/01/26 18:16:41 Success to post log:  270000
2017/01/26 18:16:41 Success to post log:  280000
2017/01/26 18:16:41 Success to post log:  290000
2017/01/26 18:16:41 Success to post log:  300000
2017/01/26 18:16:41 Success to post log:  310000
2017/01/26 18:16:41 Success to post log:  320000
2017/01/26 18:16:41 Success to post log:  330000
2017/01/26 18:16:41 Success to post log:  340000
2017/01/26 18:16:41 Success to post log:  350000
2017/01/26 18:16:51 Error while posting log:  write tcp 127.0.0.1:51954->127.0.0.1:24224: i/o timeout
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Error while posting log:  fluent#send: can't send logs, client is reconnecting
2017/01/26 18:16:51 Success to post log:  360000
2017/01/26 18:16:51 Success to post log:  370000
...
```
</details>

**Comment by tagomoris** [comment]
It's reasonable for me to add a new option to set a timeout to `logger.Post`.
It should be a option, not changing default behavior.

**Comment by yoheimuta** [comment]
Thanks !
I'm going to create a PR.

**Comment by yoheimuta** [comment]
> I'm going to create a PR.

merged #45
